migrate modules used by the HLT menu to multithreading

fwyzard opened the pull request:
migrate the following modules to be `stream` or `global` modules:
- EcalRecalibRecHitProducer
- EgammaHLTNxNClusterProducer
- HLTRegionalEcalResonanceFilter
- IPTCorrector
- IsolatedEcalPixelTrackCandidateProducer
- IsolatedPixelTrackCandidateProducer
- L1GlobalTrigger
- L1TCaloUpgradeToGCTConverter
- LaserAlignmentEventFilter
- TriggerResultsFilter
- TriggerResultsFilterFromDB

fwyzard committed:
LaserAlignmentEventFilter: change into a global::EDFilter

  - change all configuration parameters into const members
  - initialise them in the member initializer list
  - remove unnecessary parameters
  - remove unnecessary caching of EventSetup products
  - remove unused methods and member variables
  - remove usage of exceptions in the FED check loop
  - change into a global::EDFilter

fwyzard committed:
EgammaHLTNxNClusterProducer: change into a stream::EDProducer

fwyzard committed:
EcalRecalibRecHitProducer: change into a global::EDProducer

  - change all configuration parameters into const members
  - initialise them in the member initializer list
  - remove unused methods and member variables
  - change into a global::EDProducer
  - make the consumes conditional on the data actually being used

fwyzard committed:
L1TCaloUpgradeToGCTConverter: change into a global::EDProducer

  - change all configuration parameters into const members
  - initialise them in the member initializer list
  - remove unused methods and member variables
  - remove "using namespace" from header file
  - change into a global::EDProducer

fwyzard committed:
IPTCorrector: change into a global::EDProducer

  - change all configuration parameters into const members
  - initialise them in the member initializer list
  - remove unused methods and member variables
  - change into a global::EDProducer

fwyzard committed:
IsolatedPixelTrackCandidateProducer: change into a stream::EDProducer

  - change most configuration parameters into const members
  - initialise them in the member initializer list
  - change into a stream::EDProducer

fwyzard committed:
IsolatedEcalPixelTrackCandidateProducer: change into a global::EDProducer

  - change all configuration parameters into const members
  - initialise them in the member initializer list
  - remove unused methods
  - change into a global::EDProducer

fwyzard committed:
L1GlobalTrigger: change into a stream::EDProducer

  - change most configuration parameters into const members
  - change into a stream::EDProducer

fwyzard committed:
HLTRegionalEcalResonanceFilter: change into a edm::stream::EDFilter

fwyzard committed:
TriggerResultsFilter: change into a edm::stream::EDFilter

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by @fwyzard (Andrea Bocci) for CMSSW_7_6_X.

migrate to modules used by the HLT menu multithreading

It involves the following packages:

Alignment/LaserAlignment
Calibration/HcalIsolatedTrackReco
HLTrigger/HLTfilters
HLTrigger/special
L1Trigger/GlobalTrigger
L1Trigger/L1TCalorimeter
RecoEgamma/EgammaHLTProducers
RecoLocalCalo/EcalRecProducers
RecoLocalMuon/RPCRecHit


@perrotta, @cmsbuild, @diguida, @cvuosalo, @fwyzard, @cerminar, @Martin-Grunewald, @slava77, @mmusich, @mulhearn can you please review it and eventually sign? Thanks.
@ghellwig, @Sam-Harper, @jhgoh, @argiro, @Martin-Grunewald, @bellan, @mschrode, @lgray this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented on RecoLocalMuon/RPCRecHit/src/RPCPointProducer.cc:
```diff
@@ -25,44 +25,27 @@
-  clock_gettime(CLOCK_REALTIME, &start_time);  
-  */
-
+void RPCPointProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
```
this looks way too simple compared to the failed attempt in https://github.com/cms-sw/cmssw/pull/9760
DTSegtoRPC etc are using global (static) cache variables

fwyzard replied:
Thanks for pointing it out.
Indeed I have checked for static analysis reports only for the modules themselves, not for the other code that they use.
Is there a simple way to find all thread-safety issues related to a plugin ?

slava77 replied:
I know only of running helgrind
@Dr15Jones may know better

lgray replied:
Best way to do it is to walk yourself through the code in CMSSW, lets you
catch everything but externals with problems and then really complicated
template code like stuff in reco::parser.

helgrind works but it is a bit noisey and things all of the std::atomic<>
variables are causing data races, hard to figure out what exactly is
happening.

On Mon, Aug 24, 2015 at 1:57 AM, Slava Krutelyov <notifications@github.com>
wrote:

> In RecoLocalMuon/RPCRecHit/src/RPCPointProducer.cc
> <https://github.com/cms-sw/cmssw/pull/10909#discussion_r37712383>:
>
> > -
> > -RPCPointProducer::~RPCPointProducer(){
> > -
> > -}
> > -
> > -void RPCPointProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup){
> > -  /*
> > -  struct timespec start_time, stop_time;
> > -  time_t fs;
> > -  time_t fn;
> > -  time_t ls;
> > -  time_t ln;
> > -  clock_gettime(CLOCK_REALTIME, &start_time);
> > -  */
> > -
> > +void RPCPointProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
>
> I know only of running helgrind
> @Dr15Jones <https://github.com/Dr15Jones> may know better
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10909/files#r37712383>.
>

cmsbuild commented:
-1 
Tested at: cf42e6037d6f5c4dad49ae3ca00f31198b658add
 When I ran the RelVals I found an error in the following worklfows: 
 4.22 step1
<pre>DAS Error</pre>
1000.0 step1
<pre>DAS Error</pre>
1001.0 step1
<pre>DAS Error</pre>
1003.0 step1
<pre>DAS Error</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7559/summary.html

fwyzard commented:
I've dropped  the migration of RPCPointProducer for the time being.

cmsbuild commented:
Pull request #10909 was updated. @perrotta, @cmsbuild, @diguida, @cvuosalo, @fwyzard, @cerminar, @Martin-Grunewald, @slava77, @mmusich, @mulhearn can you please check and sign again.

Dr15Jones commented:
Look at http://cms-sw.github.io/showIB.html and click on the link 'Modules to thread unsafe statics'. The link 'Modules to thread unsafe EventSetup products' isn't accurate because some safe EventSetup products are falsely marked

Dr15Jones commented on Calibration/HcalIsolatedTrackReco/interface/IsolatedEcalPixelTrackCandidateProducer.h:
```diff
@@ -25,16 +25,17 @@ class IsolatedEcalPixelTrackCandidateProducer : public edm::EDProducer {
-  edm::EDGetTokenT<EcalRecHitCollection> tok_ee;
-  edm::EDGetTokenT<EcalRecHitCollection> tok_eb;
-  edm::EDGetTokenT<trigger::TriggerFilterObjectWithRefs> tok_trigcand;
+  virtual void beginJob() override;
```
The beingJob and endJob methods are empty, why not just get rid of them?

Dr15Jones commented on RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc:
```diff
@@ -31,37 +31,26 @@
+EcalRecalibRecHitProducer::EcalRecalibRecHitProducer(const edm::ParameterSet& ps) :
+   EBRecHitCollection_(        ps.getParameter<edm::InputTag>("EBRecHitCollection") ),
+   EERecHitCollection_(        ps.getParameter<edm::InputTag>("EERecHitCollection") ),
+   EBRecHitToken_(             consumes<EBRecHitCollection>(EBRecHitCollection_) ),
```
These consumes are actually conditional. The getByToken call is only made if the label of the InputTag is not null.

fwyzard replied:
Ehm, where do I see that ?

Dr15Jones replied:
line 62 and 66

fwyzard replied:
Ah, I see - sorry I misunderstood your comment before.

So, his pull request does not change the present behaviour - it was already wrong.
Or course I can add a fix for that.

Dr15Jones commented:
From a cursory look, everything seems fine. I also couldn't find any of the modified modules in the most recent 'module to statics' report:

https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-static-analysis/CMSSW_7_6_X_2015-08-23-2300/slc6_amd64_gcc493/reports/modules2statics.txt

fwyzard commented:
updated to address Chris comments

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10909 was updated. @perrotta, @cmsbuild, @diguida, @cvuosalo, @fwyzard, @cerminar, @Martin-Grunewald, @slava77, @mmusich, @mulhearn can you please check and sign again.

cmsbuild commented:
+1
Tested at: f417e43c40ea48910aede76b00927939d4aa6507
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7588/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7588/summary.html

fwyzard commented:
-1

fwyzard commented:
more work is needed for the modules that read the L1GtTriggerMenu from the EventSetup

fwyzard committed:
keep a local copy of the L1GtTriggerMenu

make a local copy of the L1GtTriggerMenu instead of modifying the one in the EventSetup
use an std::unique_ptr to manage the ownership of the local copy of the L1GtTriggerMenu

slava77 commented:
-1
? apparently this is superseded by #10927 
Andrea, please confirm

fwyzard commented:
#10927 also needs to add the new ESProducers to all configurations that make use of the RPC stuff there.
So, I'm keeping them separate for the moment, until either we find all the python snippets where the ESProducers should be added, or find an alternative approach.

fwyzard committed:
cleanup some const_cast

  - L1GtVhdlWriterCore: get rid of an unnecessary const_cast
  - L1GtTriggerMenuXmlParser: fix the input parameter of strtol

fwyzard commented:
@cmsbuild please test

fwyzard commented:
@slava77 I've removed these commits from #10927  , they can be tested and approved independently

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10909 was updated. @perrotta, @cmsbuild, @diguida, @cvuosalo, @fwyzard, @cerminar, @Martin-Grunewald, @slava77, @mmusich, @mulhearn can you please check and sign again.

fwyzard commented:
+1

fwyzard commented:
tested on workflow 1325.0 (13 TeV ttbar fullsim) with 4 threads in parallel

cmsbuild commented:
+1
Tested at: 67f77014de7db1e978026085e7aca31bfc68a5de
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7623/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7623/summary.html

slava77 commented on RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc:
```diff
@@ -31,37 +31,26 @@
+   EBRecHitCollection_(        ps.getParameter<edm::InputTag>("EBRecHitCollection") ),
+   EERecHitCollection_(        ps.getParameter<edm::InputTag>("EERecHitCollection") ),
+   EBRecHitToken_(             (not EBRecHitCollection_.label().empty()) ? consumes<EBRecHitCollection>(EBRecHitCollection_) : edm::EDGetTokenT<EBRecHitCollection>() ),
+   EERecHitToken_(             (not EERecHitCollection_.label().empty()) ? consumes<EBRecHitCollection>(EERecHitCollection_) : edm::EDGetTokenT<EERecHitCollection>() ),
```
should semantically be ``` consumes<EERecHitCollection> ```; OK since it resolves to the same type
```c++
typedef EcalRecHitCollection EBRecHitCollection;
typedef EcalRecHitCollection EERecHitCollection;
```

fwyzard replied:
sorry, copy-paste typo

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/10909	67f7701
- changes in reco related code are only in  RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer; it looks ~OK.
- jenkins tests pass and comparisons with baseline in DQM or reco fwlite comparisons show no differences as expected

cmsbuild commented:
Pull request #10909 was updated. @perrotta, @cmsbuild, @diguida, @cvuosalo, @fwyzard, @cerminar, @Martin-Grunewald, @slava77, @mmusich, @mulhearn can you please check and sign again.

slava77 commented:
uhm, a very timely rebase (just seconds after I signed)

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented:
well, I did the rebase simply to fix the typo you spotted, do we really need to test it all again ?

slava77 commented:
@fwyzard was it possible without a rebase?
I have no idea now what you changed (not from the web interface)

fwyzard commented:
```
$ git --no-pager diff --color 67f7701 2259c5d
diff --git a/RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc b/RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc
index e9fd90f..6b293d5 100644
--- a/RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc
+++ b/RecoLocalCalo/EcalRecProducers/plugins/EcalRecalibRecHitProducer.cc
@@ -35,7 +35,7 @@ EcalRecalibRecHitProducer::EcalRecalibRecHitProducer(const edm::ParameterSet& ps
    EBRecHitCollection_(        ps.getParameter<edm::InputTag>("EBRecHitCollection") ),
    EERecHitCollection_(        ps.getParameter<edm::InputTag>("EERecHitCollection") ),
    EBRecHitToken_(             (not EBRecHitCollection_.label().empty()) ? consumes<EBRecHitCollection>(EBRecHitCollection_) : edm::EDGetTokenT<EBRecHitCollection>() ),
-   EERecHitToken_(             (not EERecHitCollection_.label().empty()) ? consumes<EBRecHitCollection>(EERecHitCollection_) : edm::EDGetTokenT<EERecHitCollection>() ),
+   EERecHitToken_(             (not EERecHitCollection_.label().empty()) ? consumes<EERecHitCollection>(EERecHitCollection_) : edm::EDGetTokenT<EERecHitCollection>() ),
    EBRecalibRecHitCollection_( ps.getParameter<std::string>("EBRecalibRecHitCollection") ),
    EERecalibRecHitCollection_( ps.getParameter<std::string>("EERecalibRecHitCollection") ),
    doEnergyScale_(             ps.getParameter<bool>("doEnergyScale") ),
```

fwyzard commented:
By the way, if you prefer I can make a separate PR with only the RECO changes ?

slava77 commented:
no particular preference, I can figure out which files are RECO-related.

cmsbuild commented:
+1
Tested at: 2259c5d0a1c067435e7d5c19038378ecf761f07c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7665/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10909/7665/summary.html

slava77 commented:
-1
superseded by a bunch.
Is there going to be a bunch in 75X as well?

Andrea, please close this one (or the others) so that we don't have duplicated features

fwyzard commented:
OK, I've opened #10964 to keep track of all the related pull requests.
